opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -208,6 +208,27 @@ def set_attribute(self, key: str, value: types.AttributeValue) -> None:
         if has_ended:
             logger.warning("Setting attribute on ended span.")
             return
+        
+        if not isinstance(value, (int, float, bool, str, list)):
+            logger.warning("Invalid data type for attribute value")
+            return
+        # If value is of type list, validate all items are of same,
+        # valid data type
+        if isinstance(value, list) and len(value) > 0:
+            first_element_data_type = type(value[0])
+            # If first value is numeric, validate all values are numeric
+            if first_element_data_type in (int, float):
+                if any(not isinstance(item, (int, float)) for item in value):
+                    logger.warning("All data types of attribute value arrays must be identical")
+                    return
+            # Else, validate all values are of same type
+            elif first_element_data_type in (bool, str):
+                if any(not isinstance(item, first_element_data_type) for item in value):
+                    logger.warning("All data types of attribute value arrays must be identical")
+                    return
+            else:
+                logger.warning("Invalid data type in value array")
```

Comment:
Let's define an invalid type as any type that is not `Number` (see below), `bool` or `str`. If a invalid typed element is found in the first place of the sequence, it will be reported as `Invalid data type in value array`. But if an invalid typed element is found in any other position, it will be reported as `All data types of attribute value arrays must be identical`. I think this is kinda inconsistent.

May I recommend something like this instead:

```python
from numbers import Number


def check_sequence(sequence):

    first_element_type = type(sequence[0])

    if issubclass(type(sequence[0]), Number):
        first_element_type = Number

    for element in sequence:

        if not isinstance(element, (bool, str, Number)):
            return 'invalid type'

        if not isinstance(element, first_element_type):
            return 'different type'

    return None


assert 'different type' == check_sequence([1, 2, 3.4, 'ss', 4])
assert 'invalid type' == check_sequence([1, 2, 3.4, dict(), 4])
assert None is check_sequence([1, 2, 3.4, 5])
assert None is check_sequence(['ss', 'dw', 'fw'])
assert 'different type' == check_sequence(['sw', 'lf', 3.4, 'ss'])
```

Reply:
Thanks for your feedback @ocelotl. 

I wasn't familiar with `Number`. That makes this much cleaner. `check_sequence` doesn't seem like a method we would want to expose for clients to use.  What is the best practice for adding this functionality somewhere? Add a `_` prefix to indicate it's private? Put this method in a `util` class?